onadata/libs/utils/csv_import.py

```diff
@@ -470,3 +366,239 @@ def submission_xls_to_csv(xls_file):
+            # remove from the missing list
+            missing_col = [x for x in missing_col if not x.startswith(col)]
+            mutliple_select_col.append(col)
+
+    if missing_col:
+        return {
+            'error_msg': (
+                u"Sorry uploaded file does not match the form. "
+                u"The file is missing the column(s): "
+                u"{0}.".format(', '.join(missing_col))
+            ),
+            'valid': False}
+
+    # Identify any additional columns between XForm and
+    # imported CSV ignoring repeat and multiple_select columns
+    additional_col = [
+        col for col in set(csv_headers).difference(xform_headers)
+        if col.find('[') == -1 and col not in mutliple_select_col]
+
+    x_json = json.loads(xform.json)
+
+    # Retrieve the columns we should validate values for
+    # Currently validating date, datetime, integer and decimal columns
+    columns = {
+        'date': (get_columns_by_type(XLS_DATE_FIELDS, x_json), parse),
+        'datetime': (get_columns_by_type(XLS_DATETIME_FIELDS, x_json), parse),
+        'integer': (get_columns_by_type(['integer'], x_json), int),
+        'decimal': (get_columns_by_type(['decimal'], x_json), float)
+    }
+
+    validated_rows = []
+    errors = {}
+
+    for row_no, row in enumerate(csv_reader):
+        # Remove additional columns
+        for index in additional_col:
+            del row[index]
+
+        # Remove 'n/a' values and '' values from csv
+        row = {k: v for (k, v) in row.items() if v not in ['n/a', '']}
+
+        # Validate that row data
+        row, error = validate_row(row, columns)
+
+        if error:
+            errors[row_no] = error
+
+        if not errors:
+            location_data = {}
+
+            for key in list(row):
+                # Collect row location data into separate location_data dict
+                if key.endswith(('.latitude', '.longitude', '.altitude',
+                                '.precision')):
+                    location_key, location_prop = key.rsplit(u'.', 1)
+                    location_data.setdefault(location_key, {}).update({
+                        location_prop:
+                        row.get(key, '0')
+                    })
+
+            # collect all location K-V pairs into single geopoint field(s)
+            # in location_data dict
+            for location_key in list(location_data):
+                location_data.update({
+                    location_key:
+                    (u'%(latitude)s %(longitude)s '
+                        '%(altitude)s %(precision)s') % defaultdict(
+                        lambda: '', location_data.get(location_key))
+                })
+
+            row = csv_dict_to_nested_dict(row)
+            location_data = csv_dict_to_nested_dict(location_data)
+            row = dict_merge(row, location_data)
+
+            validated_rows.append(row)
+
+    return {
+        'valid': True if not errors else False,
```

Comment:
wouldn't `not errors` give you the desired output?

Reply:
Yes, it would. I've changed this to `not errors`